src/main/java/com/amazon/dlic/auth/http/jwt/keybyoidc/HTTPOpenIdAuthenticator.java

```diff
@@ -0,0 +1,289 @@
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import org.opensearch.OpenSearchSecurityException;
+import org.opensearch.common.settings.Settings;
+import org.opensearch.common.util.concurrent.ThreadContext;
+import org.opensearch.security.auth.HTTPAuthenticator;
+import org.opensearch.security.filter.SecurityRequest;
+import org.opensearch.security.filter.SecurityResponse;
+import org.opensearch.security.user.AuthCredentials;
+
+import com.amazon.dlic.auth.http.jwt.AbstractHTTPJwtAuthenticator;
+import com.amazon.dlic.util.SettingsBasedSSLConfigurator;
+import com.nimbusds.common.contenttype.ContentType;
+import com.nimbusds.jwt.JWTClaimsSet;
+import com.nimbusds.jwt.SignedJWT;
+import com.nimbusds.oauth2.sdk.http.HTTPRequest;
+import com.nimbusds.oauth2.sdk.http.HTTPResponse;
+import com.nimbusds.oauth2.sdk.token.AccessToken;
+import com.nimbusds.oauth2.sdk.token.AccessTokenType;
+import com.nimbusds.oauth2.sdk.util.StringUtils;
+import com.nimbusds.openid.connect.sdk.UserInfoRequest;
+import com.nimbusds.openid.connect.sdk.UserInfoResponse;
+
+import static org.apache.hc.core5.http.HttpHeaders.AUTHORIZATION;
+import static com.amazon.dlic.auth.http.jwt.keybyoidc.OpenIdConstants.CLIENT_ID;
+import static com.amazon.dlic.auth.http.jwt.keybyoidc.OpenIdConstants.ISSUER_ID_URL;
+import static com.amazon.dlic.auth.http.jwt.keybyoidc.OpenIdConstants.SUB_CLAIM;
+
+public class HTTPOpenIdAuthenticator implements HTTPAuthenticator {
+
+    private final static Logger log = LogManager.getLogger(HTTPOpenIdAuthenticator.class);
+    private final Settings settings;
+    private final Path configPath;
+    private final int requestTimeoutMs = 10000;
+    private final SettingsBasedSSLConfigurator.SSLConfig sslConfig;
+    private final String userInfoEndpoint;
+    private volatile HTTPJwtKeyByOpenIdConnectAuthenticator openIdJwtAuthenticator;
+
+    public HTTPOpenIdAuthenticator(Settings settings, Path configPath) throws Exception {
+        this.settings = settings;
+        this.configPath = configPath;
+        this.sslConfig = getSSLConfig(settings, configPath);
+        userInfoEndpoint = settings.get("userinfo_endpoint");
+        openIdJwtAuthenticator = new HTTPJwtKeyByOpenIdConnectAuthenticator(settings, configPath);
+    }
+
+    public HTTPJwtKeyByOpenIdConnectAuthenticator getOpenIdJwtAuthenticator() {
+        return openIdJwtAuthenticator;
+    }
+
+    public String getType() {
+        return "oidc";
+    }
+
+    private static SettingsBasedSSLConfigurator.SSLConfig getSSLConfig(Settings settings, Path configPath) throws Exception {
+        return new SettingsBasedSSLConfigurator(settings, configPath, "openid_connect_idp").buildSSLConfig();
+    }
+
+    @Override
+    public AuthCredentials extractCredentials(SecurityRequest request, ThreadContext context) throws OpenSearchSecurityException {
+        if (this.userInfoEndpoint != null && !this.userInfoEndpoint.isBlank()) {
```

Comment:
wdyt about moving the logic that calls the userinfo endpoint into a new method?

In AbstractHttpJwtAuthenticator would it make sense to add a new method called:

```
public JwtClaimsSet(String jwt) { 
   return jwt.getJWTClaimsSet(); // default implementation
}
```

That way this line can be replaced: https://github.com/opensearch-project/security/blob/main/src/main/java/com/amazon/dlic/auth/http/jwt/AbstractHTTPJwtAuthenticator.java#L131

In the OpenID authenticator can we then subclass that and either return `jwt.getJWTClaimsSet()` (if userinfo endpoint is not configured) or if userinfo is configured then follow the logic that you have below.

Reply:
I am not sure what the benefit of further moving things would be. We already have separation of the userinfo vs. non-userinfo logic inside this handler. If you really prefer it that way, I can move things around, but don't see how 


```
   @Override
    public AuthCredentials extractCredentials(SecurityRequest request, ThreadContext context) throws OpenSearchSecurityException {
        return new AuthCredentials(this.getJwtClaims())
    }
     
    public JwtClaimsSet getJwtClaims() {
     if (this.userInfoEndpoint != null && !this.userInfoEndpoint.isBlank()) {
            return extractCredentials0(request, context);
        }
        return (this.openIdJwtAuthenticator.extractCredentials(request, context));
   }
```
is better than 
```
   @Override
    public AuthCredentials extractCredentials(SecurityRequest request, ThreadContext context) throws OpenSearchSecurityException {
        if (this.userInfoEndpoint != null && !this.userInfoEndpoint.isBlank()) {
            return extractCredentials0(request, context);
        }
        return (this.openIdJwtAuthenticator.extractCredentials(request, context));
    }
```


These are functionally identical in my opinion and make the code harder to follow I feel. You can decide what you prefer though.

Reply:
This class would not need to override `extractCredentials`, it would completely re-use `extractCredentials` from AbstractHTTPJwtAuthenticator, but override the implementation of `getJWTClaimsSet`.